README.md
<pre>     
                     _                _      _               _ _             _        _     _                         _                           _           _   
                    | |              (_)    | |             | (_)           | |      | |   (_)                       | |                         | |         | |  
 _ __ ___  __ _  ___| |_ ________   ___ _ __| |_ _   _  __ _| |_ _______  __| |______| |__  _  ___ _ __ __ _ _ __ ___| |__  _   _ ______ ___  ___| | ___  ___| |_ 
| '__/ _ \/ _` |/ __| __|______\ \ / / | '__| __| | | |/ _` | | |_  / _ \/ _` |______| '_ \| |/ _ \ '__/ _` | '__/ __| '_ \| | | |______/ __|/ _ \ |/ _ \/ __| __|
| | |  __/ (_| | (__| |_        \ V /| | |  | |_| |_| | (_| | | |/ /  __/ (_| |      | | | | |  __/ | | (_| | | | (__| | | | |_| |      \__ \  __/ |  __/ (__| |_ 
|_|  \___|\__,_|\___|\__|        \_/ |_|_|   \__|\__,_|\__,_|_|_/___\___|\__,_|      |_| |_|_|\___|_|  \__,_|_|  \___|_| |_|\__, |      |___/\___|_|\___|\___|\__|
                                                                                                                             __/ |                                
                                                                                                                            |___/                                
</pre>

English | [简体中文](./README-zh_CN.md)

# React Virtualized Select

![React Virtualized Select](./gif.gif)

## Getting started

Install `react-virtualized-hierarchy-select` using npm.

```shell
npm install react-virtualized-hierarchy-select --save
```

## React Virtualized HierarchySelect Props

| Name          | Type                                         | Description                                                |
| ------------- | -------------------------------------------- | ---------------------------------------------------------- |
| data          | Array:[{id:x,name:x,children:[id:x,name:x]}] | tree data                                                  |
| dataMap       | Object:{x:{id:'x',name:'x'}}                 | tree map data using the example recursive function         |
| onlyCheckLeaf | Boolean                                      | onlyCheckLeaf ? only use children Key : use parentNode key |
| checkedKeys   | Array: [id,id]                               | checked treeNode key                                       |
| onChange      | (checkedKeys) => void                        | onChange event                                             |

## Example

```
import React from "react";
import VirtualizedHierarchySelect from "react-virtualized-hierarchy-select";
import "react-virtualized-hierarchy-select/dist/index.css";

const data = [
  {
    id: 1,
    name: 1,
    children: [
      {
        id: 11,
        name: 11,
        children: [
          {
            id: 111,
            name: 111,
            children: [
              {
                id: 1111,
                name: 1111,
              },
              {
                id: 1112,
                name: 1112,
              },
            ],
          },
        ],
      },
      {
        id: 112,
        name: 112,
        children: [
          {
            id: 1121,
            name: 1121,
            children: [
              {
                id: 11211,
                name: 11211,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    id: 2,
    name: 2,
    children: [
      { id: 22, name: 22 },
      { id: 222, name: 222 },
    ],
  },
];

/**
 * 递归生成{id:obj} map集合
 * @param {Array<Object>} list 数组 []
 * @param {string} childrenkey 子类key 默认"children"
 * @param {string} keyAlias id key 默认"id"
 * @returns {object} id:object键值映射
 **/
export const recursive = (
  list = [],
  childrenkey = "children",
  keyAlias = "id"
) => {
  const map = {};
  const fn = (list) => {
    if (list.length > 0) {
      list.map((item) => {
        map[item[keyAlias]] = item;
        if (item[childrenkey] && item[childrenkey].length > 0) {
          fn(item[childrenkey]);
        }
      });
    }
  };
  fn(list);
  return map;
};

const map = recursive(data)

export default class App extends React.Component {
  state = {
    checkedKeys: []
  };

  render() {
    const { checkedKeys } = this.state;

    return (
      <VirtualizedHierarchySelect
        title="测试分类"
        data={data}
        dataMap={map}
        checkedKeys={this.state.checkedKeys}
        onChange={(e) => this.setState({ checkedKeys: e })}
      />
    );
  }
}

```

## Donate

> contact me

## Contact me

> 1051919278@qq.com


CHANGELOG.md
# Changelog

### 1.0.0 (Nov 3, 2019)

- Init project

### 1.0.1 (Nov 31, 2019)

- Bugfix: fix pack

### 1.0.2 (Nov 3, 2019)

- Bugfix: fix pack

### 1.0.3 (Nov 3, 2020)

- feat: add Readme.md 
- Bugfix: fix pack

### 2.0.0 (Sep 1, 2020)

- feat: reset 
- Bugfix: fix some problems

### 3.0.0 (Sep 30, 2020)

- feat: checked style 
- Bugfix: 1.fix checked bug
          2.fix css style display bug

### 3.1.0 (Oct 9, 2020)
- Bugfix: 1.fix dischecked bug
          2.fix indeterminate bug
          3.remove mock data

### 3.2.0 (Oct 9, 2020)
- Bugfix: 1.fix css bug

### 3.3.0 (Oct 9, 2020)
- Bugfix: 1.fix css bug

### 3.3.0 (Oct 9, 2020)
- feat: 1.gif

babel.config.js
"use strict";

module.exports = {
  presets: [
    [
      "@babel/preset-env",
      {
        modules: false,
        targets: {
          browsers: ["last 2 versions", "ie >= 9"],
        },
      },
    ],
    "@babel/preset-react",
  ],
  plugins: [
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-proposal-object-rest-spread",
    "@babel/plugin-transform-async-to-generator",
    ["@babel/plugin-proposal-decorators", { legacy: true }],
    "@babel/plugin-transform-runtime",
    ["@babel/plugin-proposal-class-properties", { loose: true }],
    "@babel/plugin-proposal-export-default-from",
    "@babel/plugin-proposal-logical-assignment-operators",
    ["@babel/plugin-proposal-optional-chaining", { loose: false }],
    ["@babel/plugin-proposal-pipeline-operator", { proposal: "minimal" }],
    ["@babel/plugin-proposal-nullish-coalescing-operator", { loose: false }],
    "@babel/plugin-proposal-do-expressions",
    "@babel/plugin-proposal-function-bind",
    ["import", { libraryName: "antd", "libraryDirectory": "es", style: true  }],
  ],
};


webpack.config.js
const path = require("path");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");

let loose = true,
  modules = false,
  useESModules = true,
  filePath = "es";

module.exports = {
  mode: "production",

  entry: {
    index: "./src/components/index.jsx",
  },

  output: {
    filename: "[name].js",
    path: path.resolve(__dirname, "dist"),
    umdNamedDefine: true, // 是否将模块名称作为 AMD 输出的命名空间
    libraryTarget: "umd",
    libraryExport: "default",
  },

  module: {
    rules: [
      {
        test: /\.js[x]?$/,
        exclude: /node_modules/,
        loader: "babel-loader",
        options: {
          cacheDirectory: true,
          presets: [
            ["@babel/preset-env", { loose, modules }],
            "@babel/preset-react",
          ],
          plugins: [
            ["@babel/plugin-proposal-decorators", { legacy: true }],
            "@babel/plugin-proposal-class-properties",
            "@babel/plugin-proposal-export-default-from",
            ["@babel/plugin-transform-runtime", { useESModules }],
            [
              "import",
              {
                libraryName: "antd",
                libraryDirectory: "es",
                style: true,
              },
            ],
          ],
        },
      },
      {
        test: /\.css$/,
        include: /src/,
        use: [MiniCssExtractPlugin.loader, "css-loader", "postcss-loader"],
      },
    ],
  },

  plugins: [
    new CleanWebpackPlugin(),
    new MiniCssExtractPlugin({
      filename: "[name].css",
      chunkFilename: "[id].css",
    }),
  ],

  externals: {
    // 定义外部依赖，避免把react和react-dom打包进去
    react: {
      root: "React",
      commonjs2: "react",
      commonjs: "react",
      amd: "react",
    },
    "react-dom": {
      root: "ReactDOM",
      commonjs2: "react-dom",
      commonjs: "react-dom",
      amd: "react-dom",
    },
    lodash: {
      root: "lodash",
      commonjs2: "lodash",
      commonjs: "lodash",
      amd: "lodash",
    },
    "react-virtualized": {
      root: "react-virtualized",
      commonjs2: "react-virtualized",
      commonjs: "react-virtualized",
      amd: "react-virtualized",
    },
  },
};


webpack.development.config.js
const path = require("path");
var webpack = require("webpack");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");

module.exports = {
  mode: "development",

  entry: {
    app: "./src/index.js",
  },

  output: {
    filename: "./dev/[name].js",
    publicPath: "",
  },

  module: {
    rules: [
      {
        test: /\.js[x]?$/,
        exclude: /node_modules/,
        loader: "babel-loader",
      },
      {
        test: /\.css$/,
        use: [MiniCssExtractPlugin.loader, "css-loader", "postcss-loader"],
      },
      {
        test: /\.less$/,
        use: [
          MiniCssExtractPlugin.loader,
          "css-loader",
          "postcss-loader",
          {
            loader: "less-loader",
            options: {
              sourceMap: true,
              javascriptEnabled: true,
            },
          },
        ],
      },
    ],
  },

  plugins: [
    new CleanWebpackPlugin(),
    new webpack.HotModuleReplacementPlugin(),
    new MiniCssExtractPlugin({
      filename: "./dev/style.css",
    }),
  ],

  devServer: {
    hot: true,
    https: true,
    compress: true,
    historyApiFallback: true,
    disableHostCheck: true,
  },
};


src/index.js
import React, { Component } from "react";
import ReactDOM from "react-dom";
import VirtualizedHierarchySelect from "./components/index.jsx";

const data = [
  {
    id: 1,
    name: 1,
    children: [
      {
        id: 11,
        name: 11,
        children: [
          {
            id: 111,
            name: 111,
            children: [
              {
                id: 1111,
                name: 1111,
              },
              {
                id: 1112,
                name: 1112,
              },
            ],
          },
        ],
      },
      {
        id: 112,
        name: 112,
        children: [
          {
            id: 1121,
            name: 1121,
            children: [
              {
                id: 11211,
                name: 11211,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    id: 2,
    name: 2,
    children: [
      { id: 22, name: 22 },
      { id: 222, name: 222 },
    ],
  },
];

/**
 * 递归生成{id:obj} map集合
 * @param {Array<Object>} list 数组 []
 * @param {string} childrenkey 子类key 默认"children"
 * @param {string} keyAlias id key 默认"id"
 * @returns {object} id:object键值映射
 **/
export const recursive = (
  list = [],
  childrenkey = "children",
  keyAlias = "id"
) => {
  const map = {};
  const fn = (list) => {
    if (list.length > 0) {
      list.map((item) => {
        map[item[keyAlias]] = item;
        if (item[childrenkey] && item[childrenkey].length > 0) {
          fn(item[childrenkey]);
        }
      });
    }
  };
  fn(list);
  return map;
};
const map = recursive(data);
class Test extends Component {
  state = {
    checkedKeys: [],
  };

  render() {
    return (
      <VirtualizedHierarchySelect
        title="测试分类"
        data={data}
        dataMap={map}
        checkedKeys={this.state.checkedKeys}
        onChange={(e) => this.setState({ checkedKeys: e })}
      />
    );
  }
}

ReactDOM.render(<Test />, document.querySelector("#container"));
